iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/selector/impl/SettleSelectorImpl.java

```diff
@@ -0,0 +1,328 @@
+    try {
+      for (TsFileResource resource : resources) {
+        if (resource.getStatus() != TsFileResourceStatus.NORMAL) {
+          continue;
+        }
+        boolean shouldStop = false;
+        DirtyStatus dirtyStatus;
+        if (!heavySelect) {
+          dirtyStatus = selectFileBaseOnModSize(resource);
+        } else {
+          dirtyStatus = selectFileBaseOnDirtyData(resource);
+        }
+
+        switch (dirtyStatus) {
+          case ALL_DELETED:
+            allDirtyResource.add(resource);
+            break;
+          case PARTIAL_DELETED:
+            shouldStop = partialDirtyResource.add(resource, dirtyStatus.getDirtyDataSize());
+            break;
+          case NOT_SATISFIED:
+            shouldStop = !partialDirtyResource.getResources().isEmpty();
+            break;
+          default:
+            // do nothing
+        }
+
+        if (shouldStop) {
+          partialDirtyResourceList.add(partialDirtyResource);
+          partialDirtyResource = new PartialDirtyResource();
+          if (!heavySelect) {
+            // Non-heavy selection is triggered more frequently. In order to avoid selecting too
+            // many files containing mods for compaction when the disk is insufficient, the number
+            // and size of files are limited here.
+            break;
+          }
+        }
+      }
+      partialDirtyResourceList.add(partialDirtyResource);
+      return creatTask(allDirtyResource, partialDirtyResourceList);
+    } catch (Exception e) {
+      LOGGER.error(
+          "{}-{} cannot select file for settle compaction", storageGroupName, dataRegionId, e);
+    }
+    return Collections.emptyList();
+  }
+
+  private DirtyStatus selectFileBaseOnModSize(TsFileResource resource) {
+    ModificationFile modFile = resource.getModFile();
+    if (modFile == null || !modFile.exists()) {
+      return DirtyStatus.NOT_SATISFIED;
+    }
+    return modFile.getSize() > config.getInnerCompactionTaskSelectionModsFileThreshold()
+            || (!heavySelect
+                && !CompactionUtils.isDiskHasSpace(
+                    config.getInnerCompactionTaskSelectionDiskRedundancy()))
+        ? PARTIAL_DELETED
+        : DirtyStatus.NOT_SATISFIED;
+  }
+
+  /**
+   * Only when all devices with ttl are deleted may they be selected. On the basic of the previous,
+   * only when the number of deleted devices exceeds the threshold or has expired for too long will
+   * they be selected.
+   *
+   * @return dirty status means the status of current resource.
+   */
+  private DirtyStatus selectFileBaseOnDirtyData(TsFileResource resource)
+      throws IOException, IllegalPathException {
+    ModificationFile modFile = resource.getModFile();
+    DeviceTimeIndex deviceTimeIndex =
+        resource.getTimeIndexType() == ITimeIndex.FILE_TIME_INDEX_TYPE
+            ? resource.buildDeviceTimeIndex()
+            : (DeviceTimeIndex) resource.getTimeIndex();
```

Comment:
Will DeviceTimeIndex be downgraded during this process?

Reply:
Resolved.